cql_server::connection: Process rebounce message in case of multiple shard migrations

bitpathfinder opened the pull request:
During a query execution, the query can be re-bounced to another shard if the requested data is located there. Previous implementation assumed that the shard cannot be changed after first re-bounce, however with the introduction of Tablets, data could be migrated to another shard after the query was already re-bounced, causing a failure of the query execution. To avoid this issue, the query is re-bounced as needed until it is executed on the correct shard.

Fixes #15465

avikivity commented:
> This PR inherits from #17309 "cql_server::connection: process: rebounce msg if needed" What was added in this PR (compared with PR 17309)
> 
> * Changes are rebased with the latest master
> * A test reproducing the issue was added
> * Changed `std::invoke_result_t` concept  to `std::is_invocable_r`

Please make the cover letter stand alone and not rely on the reader knowing anything about 17309.

bitpathfinder commented:
New version: Fixing a build error.

bitpathfinder commented:
> > This PR inherits from #17309 "cql_server::connection: process: rebounce msg if needed" What was added in this PR (compared with PR 17309)
> > 
> > * Changes are rebased with the latest master
> > * A test reproducing the issue was added
> > * Changed `std::invoke_result_t` concept  to `std::is_invocable_r`
> 
> Please make the cover letter stand alone and not rely on the reader knowing anything about 17309.

PR description was updated.

bitpathfinder commented:
@avikivity after discussing this item with  @kbr-scylla, we think that we should like split the fix for message multi-rebounce and selection of the shard for executing the query (as per [these](https://github.com/scylladb/scylladb/pull/17309#issuecomment-1952233999) comments in the base PR)   to different issues since these are different problems. What's your opinion on that?

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11455/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11455/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_query_rebounce
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11455/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/445/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11516/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3493/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11455/)
#### Build Details:
- Duration: 3 hr 29 min
- Builder: spider1.cloudius-systems.com

avikivity commented:
> Fix the case when a query message is re-bounced several times (for example, with tablets migration).



7 commits, 1 sentence to explain.

gleb-cloudius commented on transport/server.hh:
```diff
@@ -140,6 +141,7 @@ struct connection_service_level_params {
     qos::service_level_options::workload_type workload_type;
 };
 
+
```
Spurious empty line

bitpathfinder commented on transport/server.cc:
```diff
@@ -976,12 +976,12 @@ template <typename Process>
             [this, is = std::move(is), cs = cs.move_to_other_shard(), stream, permit = std::move(permit), process_fn,
              gt = tracing::global_trace_state_ptr(std::move(trace_state)),
-             cached_vals = std::move(bounce_msg->take_cached_pk_function_calls()), dialect] (cql_server& server) {
+             cached_vals = std::move(cached_vals), dialect] (cql_server& server) mutable {
```
Good catch. I guess the limitation is because of the single-threaded custom allocator?

gleb-cloudius replied:
> Good catch. I guess the limitation is because of the single-threaded custom allocator?

Yes. We do have cross shard free (mostly because of exception which we do not control where they are allocated), but it is less efficient (it is a single consumer multiple producers queue with all the locking that it entails).

gleb-cloudius commented on cql3/statements/modification_statement.cc:
```diff
@@ -349,6 +387,13 @@ modification_statement::execute_with_condition(query_processor& qp, service::que
     auto shard = service::storage_proxy::cas_shard(*s, token);
+
+    if (utils::get_local_injector().is_enabled("forced_bounce_to_shard_counter")) {
+        if constexpr (std::same_as<utils::error_injection_type, utils::error_injection<true>>) {
```
Why do we need this 'if'? If error_injection is disabled during compilation `is_enable` always returns false, so everything here should be compiled out anyway.

gleb-cloudius commented on cql3/statements/modification_statement.cc:
```diff
@@ -316,6 +316,44 @@ modification_statement::execute_without_condition(query_processor& qp, service::
+namespace {
+future<::shared_ptr<cql_transport::messages::result_message>>
+process_forced_rebounce(const unsigned shard, query_processor& qp, const query_options& options) {
+    static std::atomic<int64_t> counter = {0};
```
If this has to be atomic if we execute only one lwt query simultaneously?

gleb-cloudius commented on cql3/statements/modification_statement.cc:
```diff
@@ -316,6 +316,44 @@ modification_statement::execute_without_condition(query_processor& qp, service::
+
+    logger.info("Applying forced_bounce_to_shard_counter.");
+    co_return co_await make_ready_future<shared_ptr<cql_transport::messages::result_message>>(
+        qp.bounce_to_shard(shard, std::move(const_cast<cql3::query_options&>(options).take_cached_pk_function_calls())));
```
So we always bounce to the same shard? Don't we want to move between shards on each bounce?

bitpathfinder replied:
Yes, I use the same shard, since this still reproduces multiple re-bouncing issue.

bitpathfinder commented:
New version (addressing review comments):

- Removed std::atomic.
- Removed redundant `constexpr if`..
- Made test rebounce to different shards.
- Removed std::move for cached values.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11501/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11501/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_query_rebounce
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11501/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/472/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3523/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11543/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11501/)
#### Build Details:
- Duration: 3 hr 8 min
- Builder: spider3.cloudius-systems.com

gleb-cloudius commented on transport/server.cc:
```diff
@@ -976,12 +976,12 @@ template <typename Process>
+            [this, is, cs = cs.move_to_other_shard(), stream, permit = std::move(permit), process_fn,
              gt = tracing::global_trace_state_ptr(std::move(trace_state)),
-             cached_vals = std::move(bounce_msg->take_cached_pk_function_calls()), dialect] (cql_server& server) {
+             cached_vals, dialect] (cql_server& server) mutable {
```
You dropped wrong `move`. The lambda is created on the src cpu, but the `move` below in `do_with` will be executed on a dst cpu and will move pointers from src to dst cpu. You need to drop the `move` below. Also does the lambda have to be mutable?

gleb-cloudius replied:
> Hmm. does not invoke_on run the lambda on the remote cpu? if so, the lambda captures should had already been constructed at the moment of invocation?

`invoke_on` runs lambda on the remote shard, but it creates it on a local shard. So yes lambda captures are already constructed when the lambda is invoked, but they constructed on another shard.

gleb-cloudius replied:
> in my understanding the lambda capture `cached_vals` is not a ref, but a copy, created on the src cpu, from which we are moving on the dst cpu.

This is correct. But this copy contains pointers on src cpu. When we move it (as opposite to copy) on a dst cpu we copy the pointer and then we do cross shard free of the pointer. What we should do it to move into the lamda on src and copy from the lambda on dst.

bhalevy replied:
I agree with @gleb-cloudius. If it was big and we wanted to avoid a copy we could use foreign unique ptr, so it would get freed back on the originating shard, but they way it is right now, it must not be moved on a different shard.  Previously without the `mutable` specifier, std::move didn't actually move the object content because it was const in practice, so the bug was latent.  Making the lambda mutable exposes this bug.

gleb-cloudius replied:
> @gleb-cloudius thanks for the explanation, now I see. we still can have an optimization by using a reference in [this](https://github.com/scylladb/scylladb/pull/20493/commits/f4e4d6e09364b1c1988ecceb34af519dc572001e) commit since it introduced co_await.

Doe you mean we can capture by reference? Right.

bitpathfinder replied:
Yes, I meant 'capture by reference'.

gleb-cloudius replied:
Yes. We can avoid the copy.

bitpathfinder replied:
Fixed in the new version.

bitpathfinder commented:
New version: 

- Rebased to the latest master.
- Addressed review issues.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11577/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11577/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3575/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11592/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/522/)
#### Build Details:
- Duration: 1 hr 45 min
- Builder: spider1.cloudius-systems.com

avikivity commented on transport/server.cc:
```diff
@@ -975,26 +975,19 @@ template <typename Process>
+        auto trace_state = gt.get();
+        co_return co_await process_fn(client_state, server._query_processor, in, stream, _version,
+                /* FIXME */empty_service_permit(), std::move(trace_state), false, cached_vals, dialect);
     });
```
I can't see the change, since coroutinization occured concurrently.

bitpathfinder replied:
Do you mean this change should be split to 2 commits?

bitpathfinder replied:
co-routinization is in it's own commit in the new version.

avikivity commented on cql3/statements/modification_statement.cc:
```diff
@@ -316,6 +316,60 @@ modification_statement::execute_without_condition(query_processor& qp, service::
     });
 }
 
+namespace {
```
Nicer to add a black line here.

bitpathfinder replied:
Added a blank line.

avikivity commented on cql3/statements/modification_statement.cc:
```diff
@@ -316,6 +316,60 @@ modification_statement::execute_without_condition(query_processor& qp, service::
+
+    // Select a different shard to rebounce to, while counter > 1.
+    // On the last iteration, rebounce to the correct shard.
+    if (counter != 0) {
```
Please get the spelling errors silenced or they will haunt us forever

bitpathfinder replied:
Spelling errors were fixed in the new version.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11676/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11676/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_query_rebounce
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11676/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11678/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3663/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/609/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11676/)
#### Build Details:
- Duration: 6 hr 46 min
- Builder: i-046e50f0de2689a35 (m5ad.8xlarge)

bitpathfinder commented:
New version:

- fixed review issues.

avikivity commented:
> New version:
> 
> * fixed review issues.

In the future, itemize the changes. "fixed review issues" asks the reviewer to go back to the discussion and discover what the review issues were, esp. which are newer issues and which are older issues.

avikivity commented on transport/server.cc:
```diff
@@ -1027,7 +1020,7 @@ cql_server::connection::process(uint16_t stream, request_reader in, service::cli
         auto shard = (*bounce_msg)->move_to_shard().value();
         auto&& cached_vals = (*bounce_msg)->take_cached_pk_function_calls();
-        co_return co_await process_on_shard(shard, stream, is, client_state, trace_state, dialect, std::move(cached_vals), process_fn);
+        msg = co_await process_on_shard(shard, stream, is, client_state, trace_state, dialect, std::move(cached_vals), process_fn);
```
This should still be co_return, no? The while isn't there yet.

avikivity replied:
I guess it doesn't matter since we'll return it below anyway.

bitpathfinder replied:
Yes, it's returned 2 lines later so I thought it's better to have less `return`s  from the function, also this reduces the number of changes in the next commit.

bitpathfinder commented:
> > New version:
> > 
> > * fixed review issues.
> 
> In the future, itemize the changes. "fixed review issues" asks the reviewer to go back to the discussion and discover what the review issues were, esp. which are newer issues and which are older issues.

I thought as I comment on each of review threads that it was fixed/addressed(or not), it would be a repetition to mention the same items in the summary.

avikivity commented:
> > > New version:
> > > 
> > > * fixed review issues.
> > 
> > 
> > In the future, itemize the changes. "fixed review issues" asks the reviewer to go back to the discussion and discover what the review issues were, esp. which are newer issues and which are older issues.
> 
> I thought as I comment on each of review threads that it was fixed/addressed(or not), it would be a repetition to mention the same items in the summary.

These get scattered around, they're not ordered in time.
